### PR TITLE
Refactor Result methods to rely entirely on init(create:) for catching errors and value() for switching.

### PR DESCRIPTION
**Pull request by kconner**

### CODE REVIEW

File: Sources/Result.swift [hunk 1]
```diff
@@ -76,22 +76,16 @@ public enum Result<T> {
     // Otherwise, transform the success value into a new success or failure.
     public func flatMap<U>(@noescape transform: (Value) -> Result<U>) -> Result<U> {
         return tryMap { value in
-            let mappedResult = transform(value)
-            let mappedValue = try mappedResult.value()
-            return mappedValue
+            try transform(value).value()
```
Comment (andrewroycarter) [inline Sources/Result.swift]: Looks nicer this way for sure

Comment (kconner) [inline Sources/Result.swift]: I'm starting to see why haskell programmers are tempted to reduce things to one letter ;)

File: Sources/Result.swift [hunk 2]
```diff
@@ -100,33 +94,29 @@ public enum Result<T> {
...
     }
 }
 
 // From an array of Results, return one Result of an array of their values or the first failure.
 public func zipArray<T>(results: [Result<T>]) -> Result<[T]> {
-    return results.reduce(.Success([])) { (arrayResult, itemResult) in
-        return arrayResult.tryMap { array in
-            let item = try itemResult.value()
-            return (array + [item])
+    return results.reduce(Result { [] }) { arrayResult, itemResult in
```
Comment (andrewroycarter) [inline Sources/Result.swift]: so nice that you don't need the `.Success` anymore

**Comment by andrewroycarter** [comment]
amazing 💯

**Pull request merged**